From 93a344e58e0fe1ea024f6f7936fb399956f17286 Mon Sep 17 00:00:00 2001
From: Florian Pose <fp@igh.de>
Date: Fri, 24 Jan 2020 14:36:43 +0100
Subject: [PATCH 02/24] Fixed infinite waiting for SoE write response.

---
 master/fsm_soe.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/master/fsm_soe.c b/master/fsm_soe.c
index ac0f7b06..285c9408 100644
--- a/master/fsm_soe.c
+++ b/master/fsm_soe.c
@@ -1,8 +1,6 @@
 /******************************************************************************
  *
- *  $Id$
- *
- *  Copyright (C) 2006-2008  Florian Pose, Ingenieurgemeinschaft IgH
+ *  Copyright (C) 2006-2020  Florian Pose, Ingenieurgemeinschaft IgH
  *
  *  This file is part of the IgH EtherCAT Master.
  *
@@ -619,7 +617,6 @@ void ec_fsm_soe_write_next_fragment(
         ec_print_data(data, EC_SOE_SIZE + fsm->fragment_size);
     }
 
-    req->jiffies_sent = jiffies;
     fsm->state = ec_fsm_soe_write_request;
 }
 
@@ -664,6 +661,7 @@ void ec_fsm_soe_write_start(
     fsm->offset = 0;
     fsm->retries = EC_FSM_RETRIES;
     ec_fsm_soe_write_next_fragment(fsm, datagram);
+    req->jiffies_sent = jiffies;
 }
 
 /*****************************************************************************/
@@ -716,6 +714,7 @@ void ec_fsm_soe_write_request(
         // next fragment
         fsm->retries = EC_FSM_RETRIES;
         ec_fsm_soe_write_next_fragment(fsm, datagram);
+        fsm->request->jiffies_sent = jiffies;
     } else {
         // all fragments sent; query response
         fsm->jiffies_start = fsm->datagram->jiffies_sent;
-- 
2.25.1

