From 18503344554421ab30ce444bbbf2735b69a31991 Mon Sep 17 00:00:00 2001
From: Florian Pose <fp@igh.de>
Date: Fri, 3 Apr 2020 12:09:39 +0200
Subject: [PATCH 05/24] Fixed missing include for struct sched_param.

---
 master/master.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/master/master.c b/master/master.c
index 359e85a3..446cc9f1 100644
--- a/master/master.c
+++ b/master/master.c
@@ -1,8 +1,6 @@
 /******************************************************************************
  *
- *  $Id$
- *
- *  Copyright (C) 2006-2012  Florian Pose, Ingenieurgemeinschaft IgH
+ *  Copyright (C) 2006-2020  Florian Pose, Ingenieurgemeinschaft IgH
  *
  *  This file is part of the IgH EtherCAT Master.
  *
@@ -47,7 +45,8 @@
 #include <linux/vmalloc.h>
 
 #if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 11, 0)
-#include <linux/sched/types.h> // struct sched_param
+#include <uapi/linux/sched/types.h> // struct sched_param
+#include <linux/sched/types.h> // sched_setscheduler
 #include <linux/sched/signal.h> // signal_pending
 #endif
 
-- 
2.25.1

