From ac4a67b5d8ec3272e833dbe7ced54872e6f01d68 Mon Sep 17 00:00:00 2001
From: Florian Pose <fp@igh.de>
Date: Fri, 3 Apr 2020 12:09:39 +0200
Subject: [PATCH 05/24] Fixed missing include for struct sched_param.

---
 master/master.c | 15 +++++++--------
 1 file changed, 7 insertions(+), 8 deletions(-)

diff --git a/master/master.c b/master/master.c
index bbd67714..f192c4bf 100644
--- a/master/master.c
+++ b/master/master.c
@@ -1,8 +1,6 @@
 /******************************************************************************
  *
- *  $Id$
- *
- *  Copyright (C) 2006-2012  Florian Pose, Ingenieurgemeinschaft IgH
+ *  Copyright (C) 2006-2020  Florian Pose, Ingenieurgemeinschaft IgH
  *
  *  This file is part of the IgH EtherCAT Master.
  *
@@ -46,11 +44,6 @@
 #include <linux/hrtimer.h>
 #include <linux/vmalloc.h>
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 11, 0)
-#include <linux/sched/types.h> // struct sched_param
-#include <linux/sched/signal.h> // signal_pending
-#endif
-
 #include "globals.h"
 #include "slave.h"
 #include "slave_config.h"
@@ -58,8 +51,14 @@
 #include "datagram.h"
 #include "mailbox.h"
 #ifdef EC_EOE
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(4, 11, 0)
+#include <uapi/linux/sched/types.h> // struct sched_param
+#include <linux/sched/types.h> // sched_setscheduler
+#include <linux/sched/signal.h> // signal_pending
+#endif
 #include "ethernet.h"
 #endif
+
 #include "master.h"
 
 /*****************************************************************************/
-- 
2.25.1

