From e05fe0bba160a5c8589eabe5e17252e883a052f0 Mon Sep 17 00:00:00 2001
From: Florian Pose <fp@igh.de>
Date: Wed, 17 Mar 2021 11:54:07 +0100
Subject: [PATCH 09/24] Updated gitignore.

---
 .gitignore | 40 ++++++++++++++++++++++++++--------------
 1 file changed, 26 insertions(+), 14 deletions(-)

diff --git a/.gitignore b/.gitignore
index c829cfb1..c0ebf884 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,30 +1,42 @@
-*~
+**.dirstamp
+**.ko
+**.ko.cmd
+**.lo
+**.m4
+**.mod
+**.mod.c
+**.mod.cmd
+**.o
+**.o.cmd
+**.order
+**.order.cmd
+**.swp
+**.symvers.cmd
+**/.tmp_*.gcno
+**/Module.symvers
+**~
+.deps
+.libs
+ChangeLog
+Doxyfile
+Kbuild
+Makefile
+Makefile.in
 autoconf
 autom4te.cache/
-ChangeLog
 config.h
 config.h.in
 config.log
 config.status
 configure
-.deps
-Doxyfile
 ethercat.spec
 examples/dc_user/ec_dc_user_example
 examples/user/ec_user_example
-Kbuild
 lib/libethercat.la
-.libs
 libtool
-*.lo
-*.m4
-mailbox_gateway/ethercat_mbg
-Makefile
-Makefile.in
-master/.dirstamp
-*.o
-script/ethercatctl
+master/*.o.d
 script/ethercat.service
+script/ethercatctl
 script/init.d/ethercat
 stamp-h1
 tool/ethercat
-- 
2.25.1

