From 708f2215aa5cf73a0d5e49d23148c90d4c5e91fa Mon Sep 17 00:00:00 2001
From: Mark Verrijt <mark.verrijt@vectioneer.com>
Date: Wed, 14 Apr 2021 17:51:01 +0200
Subject: [PATCH 2/3] also check device name in ethercat initd script

---
 script/init.d/ethercat.in | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/script/init.d/ethercat.in b/script/init.d/ethercat.in
index ce2bd6aa..435ecc05 100755
--- a/script/init.d/ethercat.in
+++ b/script/init.d/ethercat.in
@@ -122,11 +122,17 @@ print_dead() {
 parse_mac_address() {
     if [ -z "${1}" ]; then
         MAC=""
-    elif echo ${1} | grep -qE '^([0-9a-fA-F]{2}:){5}[0-9a-fA-F]{2}$'; then
-        MAC=${1}
     else
-        echo Invalid MAC address \"${1}\" in ${ETHERCAT_CONFIG}
-        exit_fail
+        if echo ${1} | grep -qE '^([0-9a-fA-F]{2}:){5}[0-9a-fA-F]{2}$'; then
+            MAC=${1}
+        else
+            # assume it is device name
+            MAC=$(cat /sys/class/net/${1}/address 2>/dev/null)
+            if [ -z "${MAC}" ]; then
+                echo Invalid MAC address or device name \"${1}\" in ${ETHERCAT_CONFIG}
+                exit_fail
+            fi
+        fi
     fi
 }
 
-- 
2.25.1

